### PR TITLE
Integrate Get Coins API

### PR DESCRIPTION
**Pull request by zhenyapavlovsky**
## What's new
1.  Integrated Get Coins API

## Issue

[Issue](https://trello.com/c/jAQmcx3C)

## Video

https://user-images.githubusercontent.com/99541081/227999623-5c51f4d8-4fd9-4841-8f55-459cdbd6f345.mov

### CODE REVIEW

File: Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift [hunk 1]
```diff
@@ -17,25 +17,25 @@ class MainCoordinator: ObservableObject {
...
+            case .onCoinSelected(let coinName):
+                self?.routeToDetailView(coins: coinName)
             case .onSettingsSelected:
                 self?.routeToSettingsView()
             }
         }
     }
     
-    func routeToDetailView(coin: Coin) {
+    func routeToDetailView(coins: Coins.Coin) {
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift:34]:
```suggestion
    func routeToDetailView(coin: Coins.Coin) {
```

File: Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift [hunk 2]
```diff
@@ -15,20 +17,21 @@ class DetailViewModel: ObservableObject {
     
     @Published var news: [News] = []
     
-    var coin: Coin
+    var coins: Coins.Coin
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift:20]:
```suggestion
    var coin: Coins.Coin
```

File: Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift [hunk 3]
```diff
@@ -15,20 +17,21 @@ class DetailViewModel: ObservableObject {
...
+        self.coinService = coinService
     }
     
     func navigationBack() {
         onResult?(.navigationBack)
     }
+    
+    func openURL(url: String) {
+        guard let url = URL(string: url) else { return }
+        UIApplication.shared.open(url)
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift:35]: please create separate `URLManager` and move this code inside it

File: Crypto_App_iOS/Flows/Screens/Detail/DetailView.swift [hunk 4]
```diff
@@ -124,8 +121,8 @@ private extension DetailView {
     }
 }
 
-struct DetailView_Previews: PreviewProvider {
-    static var previews: some View {
-        DetailView(viewModel: .init(coin: Coin(imageName: Image.bitcoin, nameCoin: "Bitcoin", priceCoin: 21188.04), coinService: CoinService()))
-    }
-}
+//struct DetailView_Previews: PreviewProvider {
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Detail/DetailView.swift:124]: why is this commented down?

File: Crypto_App_iOS/Flows/Screens/Detail/DetailLoadingView.swift [hunk 5]
```diff
@@ -87,9 +88,8 @@ private extension DetailLoadingView {
     }
 }
 
-struct DetailLoadingView_Previews: PreviewProvider {
-    static var previews: some View {
-        DetailLoadingView(viewModel: .init(coin: Coin(imageName: Image.bitcoin, nameCoin: "Bitcoin", priceCoin: 21188.04), coinService: CoinService()))
-    }
-}
-
+//struct DetailLoadingView_Previews: PreviewProvider {
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Detail/DetailLoadingView.swift:91]: why is it commented down?

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift [hunk 6]
```diff
@@ -9,67 +9,43 @@ import SwiftUI
 
 struct HomeView: View {
     
-    @ObservedObject var viewModel: HomeViewModel
+    @ObservedObject var viewModel = HomeViewModelImpl(
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:12]: it should be injected

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift [hunk 7]
```diff
@@ -9,67 +9,43 @@ import SwiftUI
...
-                    
+                    .navigationBarTitleDisplayMode(.inline)
                 }
-                .navigationBarTitleDisplayMode(.inline)
             }
             .background(Color.backgroundColor)
-                .onAppear() {
-                    viewModel.onAppear()
-                }
+            .onAppear() {
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:40]:
```suggestion
            .onAppear(viewModel.getCoins)
```

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift [hunk 8]
```diff
@@ -89,10 +64,67 @@ private extension HomeView {
...
+                    .background(Color.carbon)
+                    .cornerRadius(15)
+                    .padding([.leading,.trailing], 20)
+                }
+                .padding(.bottom, 15)
+            }
+        }
+    }
+    
+    func createButton(coins: Coins.Coin, action: (() -> Void)?) -> some View {
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:89]:
```suggestion
    func createButton(coin: Coins.Coin, action: (() -> Void)?) -> some View {
```

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift [hunk 9]
```diff
@@ -89,10 +64,67 @@ private extension HomeView {
...
+    }
 }
 
 struct HomeView_Previews: PreviewProvider {
     static var previews: some View {
-        HomeView(viewModel: .init(coinService: CoinService()))
+        HomeView(viewModel: HomeViewModelImpl(
+            coinService: CoinServiceImpl(
+                executor: NetworkRequestExecutor()
+                )
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:126]: bad alignment

File: Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift [hunk 10]
```diff
@@ -7,28 +7,51 @@
...
     enum Result {
         case onSettingsSelected
-        case onCoinSelected(coin: Coin)
+        case onCoinSelected(coinName: Coins.Coin)
     }
     
-    @Published var coins: [Coin] = []
+    @Published var coins = [Coins.Coin]()
+    @Published var output = String()
+    @Published var loading = false
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift:21]: please use loading state instead

File: Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift [hunk 11]
```diff
@@ -7,28 +7,51 @@
...
     
     enum Result {
         case onSettingsSelected
-        case onCoinSelected(coin: Coin)
+        case onCoinSelected(coinName: Coins.Coin)
     }
     
-    @Published var coins: [Coin] = []
+    @Published var coins = [Coins.Coin]()
+    @Published var output = String()
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift:20]: remove this and use `LoadingState`

File: Crypto_App_iOS/Services/ServiceContainer.swift [hunk 12]
```diff
@@ -12,4 +12,5 @@ class ServiceContainer {
     static let shared = ServiceContainer()
     
     lazy var userService = UserServiceImpl(executor: NetworkRequestExecutor())
```
Comment (vasylnadtochii) [inline Crypto_App_iOS/Services/ServiceContainer.swift:14]: what does `userService` stand for? 
If it is not needed, then remove it

**Review by vasylnadtochii — CHANGES_REQUESTED** [review]

File: Crypto_App_iOS/Flows/Coordination/DetailCoordinator/DetailCoordinatorView.swift [hunk 13]
```diff
@@ -15,7 +15,7 @@ struct DetailCoordinatorView: View {
     var body: some View {
         NavigationView {
             ZStack {
-                DetailView(viewModel: coordinator.detailViewModel)
+            //    DetailView(viewModel: coordinator.detailViewModel)
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/DetailCoordinator/DetailCoordinatorView.swift:18]: Done

File: Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift (hunk 1 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift:34]: Done

File: Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift (hunk 3 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift:35]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 7 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:40]: Done

File: Crypto_App_iOS/Services/ServiceContainer.swift (hunk 12 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Services/ServiceContainer.swift:14]: Done

File: Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift (hunk 2 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift:20]: Done

File: Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift [hunk 14]
```diff
@@ -15,20 +17,21 @@ class DetailViewModel: ObservableObject {
...
     var onResult: ((Result) -> Void)?
-
-    init(coin: Coin, coinService: CoinService) {
-        self.coin = coin
-        self.coinService = coinService
-    }
     
-    func onAppear() {
-        self.news = coinService.getCoinDetail()
+    init(coins: Coins.Coin, coinService: CoinService) {
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Detail/DetailViewModel.swift:24]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 7 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:40]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 8 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:89]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 9 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:126]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift (hunk 10 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift:21]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift (hunk 11 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeViewModel.swift:20]: Done

File: Crypto_App_iOS/Services/ServiceContainer.swift (hunk 12 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Services/ServiceContainer.swift:14]: Done

File: Crypto_App_iOS/Flows/Screens/Detail/DetailView.swift (hunk 4 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Detail/DetailView.swift:124]: Done

File: Crypto_App_iOS/Flows/Screens/Detail/DetailLoadingView.swift (hunk 5 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Detail/DetailLoadingView.swift:91]: Done

File: Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift [hunk 15]
```diff
@@ -17,25 +17,25 @@ class MainCoordinator: ObservableObject {
     
     @Published var route: Route?
     
-    var homeViewModel: HomeViewModel
+    var homeViewModel: HomeViewModelImpl
     
-    init() {
-        homeViewModel = HomeViewModel(coinService: CoinService())
+    init(homeViewModel: HomeViewModelImpl) {
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/MainCoordinator/MainCoordinator.swift:22]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 6 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:12]: Done

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift [hunk 16]
```diff
@@ -89,10 +58,67 @@ private extension HomeView {
...
+            }
+            .frame(height: 70)
+        }
+    }
 }
 
 struct HomeView_Previews: PreviewProvider {
     static var previews: some View {
-        HomeView(viewModel: .init(coinService: CoinService()))
+        HomeView(viewModel: HomeViewModelImpl(
```
Comment (vasyl-nadtochii) [inline Crypto_App_iOS/Flows/Screens/Home/HomeView.swift:117]: move `viewModel: HomeViewModelImpl(` to next line and format whole block of code appropriately

**Review by vasyl-nadtochii — APPROVED** [review]

File: Crypto_App_iOS/Flows/Screens/Home/HomeView.swift (hunk 16 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Home/HomeView.swift]: Done

**Pull request merged**